File: PyML-0.7.9/PyML/containers/ext/SetData.cpp

# include "SetData.h"

SetData::SetData(int size, DataSet *data_) :
  DataSet(size)
{ 
    data = data_;
    cout << "constructed SetData" << endl;
}

SetData::~SetData() 
{
  cout << "in SetData::~SetData" << endl;
}

SetData::SetData(const SetData &other, 
		 const std::vector<int> &patterns) : 
  DataSet(other, patterns),
  data(other.data)
{
  for (unsigned int i = 0; i < patterns.size(); i++) {
    int p = patterns[i];
    Y[i] = other.Y[p];
    sets.push_back(other.sets[p]);
  }
  cout << "done copy\n";
  
}

double SetData::dotProduct(int i, int j)
{
  SetData *other;
  other = this;

  return dotProduct(i, j, other);
}

double SetData::dotProduct(int i, int j, DataSet *other_)
{
  SetData *other;
  other = dynamic_cast<SetData *>(other_);
  
  float value = 0;
  for (int k = 0; k < sets[i].size(); ++k) {
      for (int l = 0; l < other->sets[j].size(); ++l) {
	  value += data->kernel->eval(this->data, sets[i][k], other->sets[j][l], other->data);
      }
  }
  return value / (sets[i].size() * other->sets[j].size());

}

void SetData::show()
{
    for (int i = 0 ; i < 2; i++) {
	cout << sets[i][0] << endl;
    }
}

File: PyML-0.7.9/PyML/containers/setData.py

from PyML.utils import misc
from PyML.utils import arrayWrap
from PyML.containers.ext import csetdata
from PyML.containers.baseDatasets import WrapperDataSet, BaseDataSet
from PyML.containers.labels import Labels

"""
classes for dealing with data that is composed of pairs of simpler objects
for which a kernel is available
"""

__docformat__ = "restructuredtext en"

class SetData (WrapperDataSet, csetdata.SetData) :

    """
    DataSet container for sets of objects.

    The kernel between two sets is defined via the kernel between the
    members of the sets:
    K(s, s') = sum_{x in s, x' in s'} K'(x, x') \ (|s| |s'|)

    file format::
    
      id1 id2 ...idk label
    
    """

    isVector = False
    
    def __init__(self, arg, **args) :
        """
        :Parameters:
          - `arg` - a file name or another SetData object.
            if a file name is supplied the constructor expects a dataset
            object as a keyword argument 'data'
        :Keywords:
          - `data` - a dataset object from which the kernel between the pairs
            of patterns is derived.
          - `patterns` - patterns to copy when performing copy construction
        """

        BaseDataSet.__init__(self, arg, **args)

    def copy(self, other, patterns, deepcopy) :

        self.callCopyConstructor(other, patterns)
        self.n = len(patterns)
        self._data = other._data

    def constructFromFile(self, file_name, **args) :

        if 'data' not in args :
            raise ValueError, 'missing data object'
        self._data = args['data']

        id_dict = misc.list2dict(self._data.labels.patternID,
                                 range(len(self._data)))
        file_handle = open(file_name)
        L = []
        sets = []
        for line in file_handle :
            tokens = line.split()
            sets.append([id_dict[token] for token in tokens[:-1] ])
            L.append(tokens[-1])
        self.n = len(sets)
        self.callConstructor(len(sets))
        for s in sets :
            self.add(tuple(s))
        labels = Labels(L)
        WrapperDataSet.attachLabels(self, labels)

    def callConstructor(self, size) :
        
        csetdata.SetData.__init__(self, size, self._data.castToBase())

    def callCopyConstructor(self, other, patterns) :
        
        csetdata.SetData.__init__(self, other, patterns)

    def __len__(self) :

        return self.n
